pkg/controller/statefulset/stateful_pod_control.go

```diff
@@ -194,6 +194,10 @@ func (spc *realStatefulPodControl) createPersistentVolumeClaims(set *apps.Statef
 		case err != nil:
 			errs = append(errs, fmt.Errorf("failed to retrieve PVC %s: %s", claim.Name, err))
 			spc.recordClaimEvent("create", set, pod, &claim, err)
+		case err == nil:
+			if pvc.DeletionTimestamp != nil {
+				errs = append(errs, fmt.Errorf("pvc %s is being deleted", claim.Name))
+			}
```

Comment:
cc @janetkuo @msau42 I think we should check the API server if the `pvc.DeletionTimestamp == nil`. Otherwise, we didn't solve the cache inconsistency issue.

Reply:
@cofyc 
Does this change make sense?
https://github.com/kubernetes/kubernetes/pull/93457/files#r478778760

Reply:
no, we can first check the informer cache, then check the API server if the `pvc.DeletionTimestamp == nil`.

- if the PVC in informer cache is in terminating state, we don't need to check API server (good performance)
  - the terminating state cann't be rollbacked, so this is reliabe 
- if the PVC in informer cache is not in terminating state (`pvc.DeletionTimestamp == nil`), we need to check the API server, because it's possible there is a delay between the API server and informer cache

Reply:
@cofyc 
OK, so how about this?  looks good?
https://github.com/kubernetes/kubernetes/pull/93457#discussion_r478793874

Reply:
This looks good to me! @tnozicka what do you think?

Reply:
how does the live call help? sounds racy the same way

maybe the statefulset should reconcile PVC (not only on pod creation) but that would be a bigger change

Reply:
this can help if the PVC is always deleted before the pod

https://github.com/kubernetes/kubernetes/issues/74374#issuecomment-682024912

> maybe the statefulset should reconcile PVC (not only on pod creation) but that would be a bigger change

yes, after some investigations, we found that hitting the issue (pod has no PVC to use) is high if the PVC is deleted after the controller checks the PVC's state because the PVC can be deleted successfully if the newly created pod is not scheduled, see the discussion here: https://github.com/kubernetes/kubernetes/issues/74374 

my suggestions:

- checking the deletion timestamp (live call maybe is not needed as we will have PVC reconciliation) 
- add PVC reconciliation